### ISSUE TITLE
Saplings, butterflies, etc use too much network data

### ISSUE THREAD

**Issue by mezz** [ForestryMC/ForestryMC#561]
Broken off from this issue #552
because leaves are fixed and they were the largest impact by far.
- [x] Saplings
- [x] Butterflies 
- [ ] anything else with genome data
- [x] anything else with owner data
- [x] anything with error states (see #666)
- [x] anything sending NBT without good reason
- [x] tiles that update everything instead of just GUI
- [x] tiles that update GUI when nobody is using the GUI
- [x] tile that send GUI updates to everyone around instead of just the user

**Comment by mezz** [comment]
I will most likely need to add a method to Item in Forge to support sending less NBT for Items. 
(thanks Skyboy for helping figure this out)
`public NBTTagCompound getClientNBT(ItemStack stack)`

Currently you can either sync all of the NBT, or none of it.

**Comment by Parker8283** [comment]
Good luck getting a patch for 1.7.10 in. You'll need it. o7

**Comment by mezz** [comment]
After a short discussion with @cpw, it looks like the patch would be too complicated and invasive to add to 1.7.10. Maybe it will be possible when he does the network code for a future Minecraft version.

**Comment by mezz** [comment]
One idea I've seen tossed around is storing a hash of the genome in the NBT, and then having a separate lookup table that maps those hashes to genomes. Then item with genomes can store less data.

Any data needed on the client would still be included in NBT, so that only the server needs to deal with the hash -> genome table.

It would also be possible to migrate from old genomes once it's set up, without breaking existing items.
It would make migrating to new formats more difficult though.

Any hash collision would be a big problem, it would probably need to be a big hash to be safe.

A further optimization might be to create a unique ID for each new genome.
This would require something like a bidirectional hash map:  ID <--> genome
When you make a new genome, hash it and look to see if it has an ID. If it does, use that ID. If not, use the next available ID and store the information in the map.
Then items only need to store a pair of IDs like 0, 1, 2, 3, ... 300000, etc to NBT instead of a much larger pair of hashes.

There would be no way to detect when an ID is no longer used, so they would have to stay around forever. I don't think this is a huge problem for servers, but performance testing for saving and loading very large tables is necessary.

**Comment by leagris** [comment]
Hello James,
Given the genome is actually little data if you remove the verbose namespace and use binary (even packed binary) I am sure it could fit into a few bits/words.
The whole genome except specie witch require a lager index can be a  table of bytes TAG_Byte_Array with each byte divided into two 4 bits nibbles: Most Significant Bit of each nibble tell if dominant/recessive. Remaining 4 bits allow for 8 different levels.

Lets look at it with real data:
TAG_Byte: 1 (replace slot 0 UID0: verbose forestry.speciesMeadows)
TAG_Byte: 1 (replace slot 0 UID0: verbose forestry.speciesMeadows)

TAG_Byte_Array:
$88 (binary 1001/1001 witch replace MSB 1 dominant 001 lowest/ MSB 1 dominant 001 lowest instead of verbose forestry.speedSlowest)
$44 (binary 0100/0100 ... pairs of recessive forestry.fertilityNormal)
$99 (binary 1010/1010 .... pairs of dominant forestry.lifespanShorter)

You don't store gene names in this array.
No variable/tag names either.
Position of entry in the genome array define the kind of gene. (1st entry is speed, 2nd, fertility, 3rd lifespan...)

Then you can fit a full genome with pairs into 14 bytes.
This is less bytes than it takes to store forestry.lifespanShorter alone.

**Comment by mezz** [comment]
I think your method makes sense with some small changes:

We will need to choose an ordering for species and genes that will not break when addon mods are added or removed. I think that means keeping a map of `Integer geneEncodedID <--> String geneUID`.
This table would be much much smaller than the genome table I thought of earlier, so this method makes more sense I think.

We can pack together the `geneEncodedID`s using [varInts](https://en.wikipedia.org/wiki/Variable-length_quantity), since we don't know how many variations of each gene there might be in a world due to addons. Unsigned varint should take up 1 byte for each gene with less than 128 (2^7) types, and up to 2 bytes for genes with up to 16k (2^14) types. 

I estimate a bee genome would pack into:
species: up to 2 bytes
13 other genes: 1 byte each
so about 30 bytes for a full genome with pairs.
The position of each entry will still define the kind of gene like you suggest.

It will be bigger than 14 bytes but I think it's the only way to allow it to be safely extended. I don't want it to break if other mods want to add hundreds of species, gene variations, and effects.

**Comment by Nedelosk** [comment]
https://github.com/MinecraftForge/MinecraftForge/pull/3342

**Comment by leagris** [comment]
awesome!

**Comment by mezz** [comment]
Last time I suggested that change it did not get through, hopefully it is different this time.

**Comment by mezz** [comment]
Yay it was accepted!
There is still work to do though, the analyzer and other client-side display of information will need special packets to receive genetic information.

**Comment by Nedelosk** [comment]
@mezz Should we still do the encoding thing or is there a better way to do this ?

**Comment by leagris** [comment]
I think it would be still worth it.
Client only need to know about rendering and client-side interaction.

Saplings all renders same regardless all genes except specie.
Although even if client don't receive a full genome NBT, it needs to differentiate genome variants so it will not attempt stacking same species regardless of NBT.
How would you do this without at least sending a genome hash to compare stack.isEqual?
But a bit-packed genome data will always be shorter than a hash of all data.

I think of no case where you could avoid sending full genome. So better have it bit packed.

**Comment by mezz** [comment]
I agree that it is probably not possible to use the client-side nbt feature.

We can encode the genome to integers using a Forge Registry for registering alleles.
Then we can pack those using an array of `varInt`. This is not as efficient as real bit packing, but it allows for adding and removing mods in a world without breaking anything.
If you take it further and make one registry for each type of allele (one for `IAlleleBeeSpecies`, one for `IAlleleFloat`, etc) then each number will be very small and they should take up even less data.

I have not really had time to work on it, if you are interested please go ahead @Nedelosk and assign me for a review. We will need to be careful about backward-compatibility. 

For forward-compatibility we should also probably prefix the data with a version number. That way if any `IChromosomeType` changes, we can increase the version number and keep compatibility. I don't think we should be changing those things but maybe it will happen when we remove some deprecated stuff.

**Comment by Nedelosk** [comment]
So we would have to save the ids in a file in the "saves" folder. What about the `ItemStack`s that we already created before a player joined a world.
